### PR TITLE
Implement circuit for `ErrorOutOfGasMemoryCopy` state

### PR DESCRIPTION
**Pull request by silathdiir**
### Description

Spec Markdown PR https://github.com/scroll-tech/zkevm-specs/pull/59

#### Summary

1. Merge OOG error `ExtCodeCopy` into `MemoryCopy`.

2. Implement bus-mapping and circuit for this OOG error of `CALLDATACOPY`, `CODECOPY`, `EXTCODECOPY` and `RETURNDATACOPY` opcodes.

3. OOG error executions of `CALLDATACOPY`, `CODECOPY` and `RETURNDATACOPY` are same.

4. `EXTCODECOPY` has extra `1` stack pop, and constant gas costs are different for cold (2600) and warm (100) accounts according to EIP-2929 (could reference [go-etherum gasExtCodeCopyEIP2929 function](https://github.com/ethereum/go-ethereum/blob/master/core/vm/operations_acl.go#L116)).

5. Fix `memory_copier_gas_cost` function to not calculate memory expansion for zero copy size in gas utils of `eth-types`.

### Issue Link

Upstream issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1266

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Add new unit-test cases for this error state.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+        let dst_memory_addr = MemoryAddressGadget::construct(cb, dst_offset, copy_size);
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [dst_memory_addr.address()]);
+        let memory_copier_gas = MemoryCopierGasGadget::construct(
+            cb,
+            dst_memory_addr.length(),
+            memory_expansion.gas_cost(),
+        );
+
+        debug_assert!(
```
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs:67]: I think just comment that these three opcode has same constant gas is OK.  remove debug_assert.

**Review by DreamWuGit — APPROVED** [review]
lgtm, another thing is oog ExtCodeCopy. seems can be merged into this error gadget. WDYT ?

**Comment by silathdiir** [comment]
> lgtm, another thing is ExtCodeCopy. seems can be merged into this error gadget. WDYT ?

Yes. Suppose to extract a common oog-memory-copy gadget in this PR. And both OOGMemoryCopyGadget and OOGExtCodeCopyGadget could use this common gadget, and handle different stack-pop values (and different constant gas cost) in self-implementation. WDYT?

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs (hunk 1 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs]: Replaced `debug_assert` with a comment in [99aba46](https://github.com/scroll-tech/zkevm-circuits/pull/368/commits/99aba4648a2423bd565276a03f9d9072c2359690).

**Comment by silathdiir** [comment]
Merged OOG error `ExtCodeCopy` into `MemoryCopy` in commit [5a1694b](https://github.com/scroll-tech/zkevm-circuits/pull/368/commits/5a1694b2dd0aad8c0d0416c8ed89e638bc6a63a8).

`EXTCODECOPY` has extra `1` stack pop, and constant gas costs are different for cold (2600) and warm (100) accounts according to EIP-2929 (could reference [go-etherum gasExtCodeCopyEIP2929 function](https://github.com/ethereum/go-ethereum/blob/master/core/vm/operations_acl.go#L116)).

**Pull request merged**